src/sessions/VirtualPageLoadTimer.ts

```diff
@@ -0,0 +1,245 @@
+        const self = this;
+        return original
+            .apply(thisArg, argsArray)
+            .catch((error) => {
+                // Silently swallowing error
+            })
+            .finally(self.decrementFetchCounter());
+    };
+
+    /**
+     * Increment the fetch counter in PageManager when fetch is beginning
+     */
+    private fetchWrapper = (): ((
+        original: (input: RequestInfo, init?: RequestInit) => Promise<Response>
+    ) => (input: RequestInfo, init?: RequestInit) => Promise<Response>) => {
+        const self = this;
+        return (original: Fetch): Fetch => {
+            return function (
+                this: Fetch,
+                input: RequestInfo,
+                init?: RequestInit
+            ): Promise<Response> {
+                self.fetchCounter += 1;
+                return self.fetch(original, this, arguments, input, init);
+            };
+        };
+    };
+
+    private decrementFetchCounter() {
+        if (!this.isPageLoaded) {
+            this.latestEndTime = Date.now();
+        }
+        this.fetchCounter -= 1;
+    }
+
+    /**
+     * Checks whether the virtual page is still being loaded.
+     * If completed:
+     * (1) Clear the timers
+     * (2) Record data using NavigationEvent
+     * (3) Indicate current page has finished loading
+     */
+    private checkLoadStatus = () => {
+        if (this.ongoingRequests.size === 0 && this.fetchCounter === 0) {
+            clearInterval(this.periodicCheckerId);
+            clearTimeout(this.timeoutCheckerId);
+            this.domMutationObserver.disconnect();
+            this.recordRouteChangeNavigationEvent(this.pageManager.getPage());
+            this.periodicCheckerId = undefined;
+            this.timeoutCheckerId = undefined;
+            this.isPageLoaded = true;
+        }
+    };
+
+    /** Clears timers and disconnects observer to stop page timing. */
+    private declareTimeout = () => {
+        clearInterval(this.periodicCheckerId);
+        this.periodicCheckerId = undefined;
+        this.timeoutCheckerId = undefined;
+        this.domMutationObserver.disconnect();
+        this.isPageLoaded = true;
+    };
+
+    /** Resets periodic check timer and updates page's latestEndTime. */
+    private resetInterval = () => {
+        const page = this.pageManager.getPage();
+        this.latestEndTime = Date.now();
+        clearInterval(this.periodicCheckerId);
+        this.periodicCheckerId = setInterval(
+            this.checkLoadStatus,
+            this.config.routeChangeLoadCheckInterval
+        );
+    };
+
+    private moveItemsFromBuffer = (item: any) => {
+        this.ongoingRequests.add(item);
+    };
+
+    private recordRouteChangeNavigationEvent(page: Page) {
+        const virtualPageNavigationEvent: NavigationEvent = {
+            version: '1.0.0',
+            initiatorType: 'route_change',
+            navigationType: 'navigate',
+            startTime: page.start,
```

Comment:
issue: `page.start` is the time when `PageManager.recordPageView` is called. However, in the design we defined the start virtual page load duration as the most recent DOM click event.